core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1318,14 +1317,20 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     trace("Sending topic metadata %s and brokers %s for correlation id %d to client %s".format(completeTopicMetadata.mkString(","),
       brokers.mkString(","), request.header.correlationId, request.header.clientId))
+    val controllerId = {
+      metadataCache.getControllerId.flatMap {
+        case ZkCachedControllerId(id) => Some(id)
+        case KRaftCachedControllerId(_) => metadataCache.getRandomAliveBrokerId
```

Comment:
See my note in MetadataCache. We could simplify this to `metadataCache.getControllerNode.map(_.id)`

Reply:
I did something like this before. But got a different comment then, saying we should explicitly handle that in KafkaApis.